### DISCUSSION TITLE
Support arbitrary amount of "admin clients"

### DISCUSSION THREAD

**Discussion by danielFesenmeyer** [keycloak/keycloak#8599] [Ideas]
Currently, there are two built-in "admin clients": admin-cli and security-admin-console. For requests with one of these clients, Keycloak authorizes based on all roles of the user, not only the ones contained in the token. In fact, there are no roles in the scope (and therefore in the created tokens) of these clients. This has the advantage that tokens don't get too big to fit into a Bearer Authorization header.

Our customers would like to create their own "admin clients" in addition to the existing ones. There are several use cases, for example these ones:

- A service account client on the master realm which is used for scheduled realm management operations.
- Having multiple clients for the same purpose, in order to allow credentials rotation

What about adding an "admin client" flag to the client representation, which leads to the behaviour described above? Some places in the code could even be simplified, for example here it could just be checked for the flag instead of checking for the special client IDs: https://github.com/keycloak/keycloak/blob/7b135c4dfce3c8cd79cf515fd02f35059a21e07e/services/src/main/java/org/keycloak/services/resources/admin/permissions/MgmtPermissions.java#L102

What do you think about such a flag?
If you think it's useful, we would be happy to provide a PR.

**Comment by stianst** [comment]
I'm not keen on adding a flag like you're suggesting, as that results in a config option that makes no sense to 99.9999% of clients, so would like to have some cleaner approach on how to fill this use-case. Maybe there's also some connection here to non-admin usage as well when there are large amounts of roles.

There's perhaps some connection here with https://github.com/keycloak/keycloak/discussions/8516 since with namespaces we could add a scope of roles to the token rather than the roles themselves, which would suggest to a resource provider that it needs to lookup what roles a user has separately. With internal clients like admin APIs they can just lookup the roles for the user internally, while external clients can use the token introspection endpoint to get the actual roles.

**Comment by stianst** [comment]
@pedroigor @dgozalo FIY

**Comment by pedroigor** [comment]
@danielFesenmeyer It is not clear to me why you need to create additional clients other than using the built-in ones. Could you please, elaborate more on that?

Also for scheduled management operations, perhaps it makes more sense to have these jobs running on Keycloak itself? It should avoid additional overhead when processing HTTP requests to our APIs and blocking worker threads.

In regards to credential rotation, looks like this one could be something we could have OOTB? I don't know ...

**Reply by stianst** [reply]
> In regards to credential rotation, looks like this one could be something we could have OOTB? I don't know ...

Client secret rotation is something we will add, see https://github.com/keycloak/keycloak/discussions/8475 for more details on that

**Reply by stianst** [reply]
@pedroigor it's not the first time this has come up - Admin APIs are quite often used, and often used from confidential clients, so not sure how "using the built-in" ones applies here? The built-in ones like the one for admin UI also is tied to KC URLs.

**Reply by pedroigor** [reply]
@stianst Yeah, they are. But if you are doing client credentials to obtain tokens, you should be fine to access the Admin API even though the built-in client is tied with KC URLs?

Or these clients are alternative Admin UIs served by server-side apps doing code grant?

In theory, people should be able to have a specific admin client (`admin-cli`) on a per realm basis, if each tenant/customer has its own realm. But I guess it is not the case here, @danielFesenmeyer? Mainly because we are not good at supporting a large number of realms.

So I guess the main requirement here is to avoid sharing the built-in admin client between tenants when in a single realm?

**Reply by jbman** [reply]
@pedroigor that's true, the main requirement is not sharing built-in admin client. Our tenant/customer wants to manage multiple realms and we want to provide them with a client but don't want to share it.

To summarize all cases which are solved by the flag:
1. Use different clients for service provider operating Keycloak and service customer who is managing multiple realms
2. Different automation/management tools/UIs which should not share the same client
3. For a credential rotation, if you want to have a phase when old and new credentials are valid

**Reply by stianst** [reply]
@pedroigor sharing the clients makes no sense at all. Basically what you are advocating for is client impersonation.

**Reply by stianst** [reply]
@jbman the use-case of being able to create additional clients that can manage many realms is perfectly valid, and would be good to support properly. However, I'm not keen on having a config flag like this, and would like us to figure out a different way to achieve it.

**Reply by danielFesenmeyer** [reply]
@stianst @jbman Actually, adding a special config flag was the only idea I came up with. I was not aware what you have planned in context of Namespaced Roles (#8516) - looks promising. I like your suggestion from above: to include a scope of roles in the token (instead of all the actual roles).

**Comment by jbman** [comment]
## Client scope  _keycloak-admin-api_

How about an other option: Instead of a flag, a client scope _keycloak-admin-api_ is defined and used at the default admin clients and on other custom ones, if required.

More details:
The client scope description can tell `Scope for accessing Keycloak Admin API without that roles need to be present in access token` and `Include In Token Scope` must be ON. A client would use this scope and not the _roles_ scope.
The place which checks for the `azp `claim with `auth.getToken().getIssuedFor()` will check for this scope value instead. This doesn't require to store an additional flag or a custom claim at the token.

@stianst What do you think, better or worse than a flag? 
Anyhow, there must be some point of configuration which allows clients to be used at Admin API without all realm roles in the token.
Do you have an other point of configuration in mind?

I would be happy with any configuration as long as it is explicitly configured and explainable to an admin user. Current approach is hard to explain, as the `azp `claim check against the two default admin clients is implicit, restricted to the two default clients and hidden in the code.

**Reply by stianst** [reply]
I like that idea much better than a config option. Could we make it more generic perhaps? What about a mapper that can set a claim in the token that means the token should allow all roles the given client has a scope on without actually listing the roles? That can then be also useful for generic use-cases where the # roles are to large for the initial access token, allowing the token introspection endpoint to be used to retrieve all the roles.

**Reply by jbman** [reply]
With a single mapper all roles would be excluded from the token, while it would be much better if a token can still contain some roles and some would only be available internally or through token introspection. This would need a config option at the client scope. A flag `exclude roles from token` which needs to be read by the existing `client roles` and `realm roles` mappers so that they don't put the roles in the token. However the client scope should still end up in the token and the roles should end up to be available at the token introspection endpoint.

**Reply by stianst** [reply]
I wouldn't add an option to client scopes to configure this, so would need some other solution to that problem. I would imagine that you wouldn't us client roles or realm roles mapper in this scenario though, or they would have config options for this purpose.

**Reply by danielFesenmeyer** [reply]
Hi @jbman, @stianst. I've implemented a draft PR with a mapper approach: https://github.com/keycloak/keycloak/pull/8914 - I've also added a detailed description, with considered alternatives. Would be great to get some feedback from you.

**Reply by stianst** [reply]
@danielFesenmeyer thanks, while looking at your PR I was wondering if there's an even simpler and more generic solution to this problem, essentially just making the role mappers add to introspection endpoint, and not access token. We can discuss that through your PR though.
